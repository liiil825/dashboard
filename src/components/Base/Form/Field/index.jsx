import React from 'react';
import {
  Checkbox, Button, Input, Select
} from 'components/Base';
import wrapField from './wrapField';
import FieldGroup from './FieldGroup';

const TextArea = props => <textarea {...props} />;

export default {
  FieldGroup,
  ButtonField: wrapField(Button),
  CheckboxField: wrapField(Checkbox),
  TextField: wrapField(Input),
  TextareaField: wrapField(TextArea),
  SelectField: wrapField(Select)
};
