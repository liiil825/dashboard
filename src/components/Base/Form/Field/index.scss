@import '~scss/vars';

.field {
  display: block;
  margin-bottom: 1.25rem;
  label {
    line-height: 1.29;
  }
  label:not(:last-child) {
    margin-bottom: 0.5em;
  }
}

.horizon {
  width: 100%;
  label {
    display: inline-flex;
    justify-content: flex-end;
    align-items: center;
    min-width: $form-label-width;
    padding-right: 1rem;
  }
}
.vertical {
  label {
    display: block;
  }
}

.vertical + .vertical {
  margin-left: 1.25rem;
}
