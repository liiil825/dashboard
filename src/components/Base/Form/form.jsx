import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _ from 'lodash';

import { getFormData } from 'utils';

import styles from './index.scss';

export default class Form extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    data: PropTypes.object,
    labelType: PropTypes.oneOf(['normal', 'title']),
    layout: PropTypes.oneOf(['horizon', 'vertical', 'inline']),
    onSubmit: PropTypes.func
  };

  static defaultProps = {
    className: '',
    size: 'normal',
    labelType: 'normal',
    layout: 'horizon'
  };

  constructor(props) {
    super(props);
    this.formRef = React.createRef();
  }

  handleSubmit = e => {
    e.preventDefault();
    this.props.onSubmit(e, getFormData(this.formRef.current));
  };

  render() {
    const {
      children, layout, labelType, noPadding, ...restProps
    } = this.props;

    const formClass = classnames(
      styles.form,
      {
        [styles.noPaddingForm]: noPadding
      },
      this.props.className
    );

    const childNodes = React.Children.map(children, child => {
      if (!child) {
        return null;
      }

      const isField = _.invoke(child, 'type.displayName.includes', 'Field');
      const className = classnames(
        styles.formItem,
        _.get(child, 'props.className')
      );
      const props = {
        ...child.props,
        className
      };
      if (isField) {
        props.layout = props.layout || layout;
        props.labelType = props.labelType || labelType;
        props.size = props.size || this.props.size;
      }
      return React.cloneElement(child, props);
    });

    return (
      <form
        {...restProps}
        className={formClass}
        ref={this.formRef}
        onSubmit={this.handleSubmit}
      >
        {childNodes}
      </form>
    );
  }
}
