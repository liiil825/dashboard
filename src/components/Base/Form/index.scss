@import '~scss/vars';

.form {
  padding: 24px 24px 32px;
}
.noPaddingForm {
  padding: 0;
}

.formItem + .formItem {
  margin-top: $form-field-gap;
}

label.formItem {
  width: 100%;
  text-align: left;
}

.field {
  display: flex;
  align-items: center;
  label {
    box-sizing: border-box;
  }
  label.normal {
    @include normal-font(
      $form-field-label-line-height,
      $form-field-label-font-size
    );
  }
  label.title {
    @include title-font(
      $form-field-label-line-height,
      $form-field-label-font-size
    );
  }
}
.field + .field {
  margin-top: $form-field-gap;
}

.horizon {
  flex-flow: row wrap;
  .control {
    width: 100% - $form-label-width;
  }
  .help {
    margin-left: $form-label-width;
    margin-top: $form-field-secondary-gap;
  }
}
.vertical {
  flex-flow: column;
  align-items: flex-start;
  label:not(:last-child) {
    margin-bottom: $form-field-secondary-gap;
  }
  label {
    display: block;
  }
  .help {
    margin-top: $form-field-secondary-gap;
  }
}
.horizon,
.inline {
  label {
    display: inline-flex;
    justify-content: flex-end;
    align-items: center;
    width: $form-label-width;
    padding-right: $form-field-main-gap;
  }
  &.textareaField {
    align-items: flex-start;
    label {
      margin-top: $form-field-secondary-gap;
    }
  }
}
.inline {
  .help {
    margin-left: $form-field-secondary-gap;
  }
}

.fieldGroup {
  display: flex;

  .vertical {
    & + .vertical {
      margin-left: $form-field-gap;
      margin-top: 0;
    }

    &.checkboxField {
      margin-top: $form-field-label-height;
    }
  }
}
.fieldGroup + .fieldGroup {
  margin-top: $form-field-main-gap;
}

.help {
  @include note-font();
}
