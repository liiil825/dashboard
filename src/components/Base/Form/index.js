import Form from './form';
import Field from './Field';

export default Object.assign(Form, Field);
