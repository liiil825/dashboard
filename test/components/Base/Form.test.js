import Form from 'components/Base/Form';

const {
  FieldGroup, ButtonField, TextField, SelectField, CheckboxField
} = Form;

describe('Base/Form', () => {
  const spyOnSubmit = jest.fn();

  const props = {
    onSubmit: spyOnSubmit
  };
  const getWrapper = func => func(
      <Form {...props}>
        <FieldGroup>
          <SelectField
            label="Server protocol"
            name="protocol"
            options={[
              {
                value: 'smtp',
                label: 'SMTP'
              },
              {
                value: 'pop3',
                label: 'POP3'
              },
              {
                value: 'imap',
                label: 'IMAP'
              }
            ]}
          />
        </FieldGroup>

        <FieldGroup>
          <TextField name="email_host" placeholder="server name" />

          <TextField
            name="port"
            placeholder="1000"
            type="number"
            size="small"
          />

          <CheckboxField name="ssl_enable">
            {'SSL secure connection'}
          </CheckboxField>
        </FieldGroup>

        <FieldGroup>
          <TextField label={'Sender nickname'} name="display_sender" />
        </FieldGroup>

        <FieldGroup>
          <TextField name="email" />
          <TextField
            label="Server password"
            name="password"
            type="password"
            placeholder="**********"
          />
        </FieldGroup>
        <FieldGroup>
          <ButtonField htmlType="submit" layout="inline">
            submit
          </ButtonField>
        </FieldGroup>
      </Form>
  );
  it('basic render', () => {
    const wrapper = getWrapper(shallow);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('call submit', () => {
    const wrapper = getWrapper(mount);
    wrapper.find('[type="submit"]').simulate('submit');
    expect(spyOnSubmit).toHaveBeenCalled();
    wrapper.unmount();
  });
});
