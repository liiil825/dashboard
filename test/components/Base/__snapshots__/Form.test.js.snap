// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Form basic render 1`] = `
<form
  className="form"
  onSubmit={[Function]}
  size="normal"
>
  <FieldGroup
    className="formItem"
    key=".0"
    labelType="normal"
    layout="horizon"
    size="normal"
  >
    <Field-Select
      label="Server protocol"
      name="protocol"
      options={
        Array [
          Object {
            "label": "SMTP",
            "value": "smtp",
          },
          Object {
            "label": "POP3",
            "value": "pop3",
          },
          Object {
            "label": "IMAP",
            "value": "imap",
          },
        ]
      }
    />
  </FieldGroup>
  <FieldGroup
    className="formItem"
    key=".1"
    labelType="normal"
    layout="horizon"
    size="normal"
  >
    <Field-Input
      name="email_host"
      placeholder="server name"
    />
    <Field-Input
      name="port"
      placeholder="1000"
      size="small"
      type="number"
    />
    <Field-Checkbox
      name="ssl_enable"
    >
      SSL secure connection
    </Field-Checkbox>
  </FieldGroup>
  <FieldGroup
    className="formItem"
    key=".2"
    labelType="normal"
    layout="horizon"
    size="normal"
  >
    <Field-Input
      label="Sender nickname"
      name="display_sender"
    />
  </FieldGroup>
  <FieldGroup
    className="formItem"
    key=".3"
    labelType="normal"
    layout="horizon"
    size="normal"
  >
    <Field-Input
      name="email"
    />
    <Field-Input
      label="Server password"
      name="password"
      placeholder="**********"
      type="password"
    />
  </FieldGroup>
  <FieldGroup
    className="formItem"
    key=".4"
    labelType="normal"
    layout="horizon"
    size="normal"
  >
    <Field-Button
      htmlType="submit"
      layout="inline"
    >
      submit
    </Field-Button>
  </FieldGroup>
</form>
`;
